### DISCUSSION TITLE
RFC: Use cases for transient workers?

### DISCUSSION THREAD

**Discussion by mschubert** [mschubert/clustermq#257] [RFC]
We've had multiple mentions how `clustermq` workers are "*persistent*" and that in some situations more "*transient*" (i.e., workers that get spun up and down more frequently) can be desirable for some use cases. This was coming from especially the angle of the `drake` and `targets` packages.

The question that I'm asking myself now is: **What are these use cases?**

I see the following:

1. Computing on a graph may parallelize to 100 workers, or go through an integration ("reduce") step where no parallelism is possible
2. (I'm not sure about batch-like queries, are there any?)

And the following gradient of approaches:

1. Spin up workers once and keep all of them until all computations are complete
2. Spin up workers once and shut them down as soon as their work is complete (*this is what we are currently using*)
3. (there is probably some intermediate uses cases here)
4. Spin up a separate worker for each individual function call

The advantages and disadvantages I see for those approaches are roughly:

1. This is wasteful, as workers are keep online when they are idle (both for startup and shutdown)
2. This seems like a better alternative than 1, but workers may still be idle until they receive their instructions (I'd consider this a minor drawback, when is this an issue?)
3. ?
4. This will likely put unnecessary load on the scheduler; a rule of thumb here is that jobs should run at least 2 minutes (and individual function calls may well be faster)

So, I am looking for examples where **point 3** (or **4**) is desirable, and why it will serve a use case better than the current approach (**point 2**).

In terms of implementation, I'm considering:

* A worker API call that allows either to (a) resize the worker pool, or (b) concatenate different (potentially heterogeneous) pools of workers; when this should be done would be a decision of the API user
* others?

/cc @wlandau @pat-s @mattwarkentin

**Comment by pat-s** [comment]
Thanks @mschubert - important topic!

Reading your post I am a bit confused because of 

> Spin up workers once and shut them down as soon as their work is complete (this is what we are currently using)

This is not what I experience in practice but I am not sure if this is due to {clustermq}. I am using {clustermq} via {drake} and the following SLURM logic

> #SBATCH --array=1-{{ n_jobs }}

In this scenario, when I spin up 100 workers, some finish earlier then others. These then output the following when their job is done

> 2021-04-03 22:48:07.480188 | > WORKER_STOP (0.000s wait)
> 2021-04-03 22:48:07.488350 | shutting down worker

The worker is in an an IDLE state and blocks resources for the scheduler.
SLURM will only release the allocated resources if **all** jobs are done.
In my detailed scenario from the 100 workers I have around 5 which take a lot longer than the rest.
This is very unfortunate because I then block a lot of resources even though just 5% of them are in actual use at some point in time.

So TL,DR: is this maybe a thing I need to tell/configure in SLURM and {clustermq} is doing just fine? 
My goal would be that the scheduler releases the resources once it realizes that they are not needed anymore to compute the missing array tasks.

**Reply by wlandau** [reply]
@pat-s, that sounds specific to SLURM. In the SGE cluster I use at work, when the `targets` pipeline gets to a "shutting down worker" stage, those workers actually shut down while the remaining ones keep going if they still have work to do.

**Reply by pat-s** [reply]
That sounds promising! I haven't though of that before yet, maybe the fix is quite simple. 

I've summarized it in this [SO question](https://stackoverflow.com/questions/66943044/slurm-free-resources-of-idle-job-array-workers) after I could not find a solution in the Slurm docs. Let's see.

**Reply by pat-s** [reply]
@wlandau [FYI](https://stackoverflow.com/questions/66943044/slurm-free-resources-of-idle-job-array-workers/66945536#66945536). Let me know if I can help in any way or if I should open an issue in {drake} on GH.

**Reply by wlandau** [reply]
As you suspected, the dynamic branching cleanup step is part of the priority queue, which is preventing the shutdown of superfluous workers. This is a design issue that targets is prepared to fix but drake is not. Tracking in https://github.com/ropensci/targets/issues/398.

**Reply by mattwarkentin** [reply]
Joining in late here, but FWIW, my experience has been the same as @pat-s. I also use `slurm` and if I use `targets`/`clustermq` and request 10 workers, e.g.:
```r
tar_make_clustermq(workers = 10L)
```
Assuming all 10 workers are available, all 10 would start, but none of these resources will be relinquished until ALL `targets` jobs are done.

Is this now fixed in `targets`, @wlandau?

**Reply by wlandau** [reply]
Should be fixed now, yes.

**Reply by mattwarkentin** [reply]
Can confirm it works on my HPC with Slurm. This is excellent!

**Reply by pat-s** [reply]
@wlandau I assume there won't be a backport for {drake}, will it?

**Reply by wlandau** [reply]
I know I promised backports for some things, but I did not architect `drake` well enough to handle the kind of flexibility required in this case, especially with dynamic branching. Even with `targets`, it was very sensitive work to implement, and I would not have caught https://github.com/ropensci/targets/issues/404 if I had not been using `targets` for an actual data analysis project at work. So in spite of the inefficiency, I would prefer to leave `drake` alone so we can be more confident that it continues to orchestrate and conclude pipelines without obvious errors.

**Reply by pat-s** [reply]
Fair enough - looking foward to use {targets} soon hopefully!

**Comment by wlandau** [comment]
@mschubert, thank you _so_ much for bringing up this thread! It is such a critical issue for large `targets` pipelines!

>The question that I'm asking myself now is: What are these use cases?
>
>I see the following:
>
> 1. Computing on a graph may parallelize to 100 workers, or go through an integration ("reduce") step where no parallelism is possible
> 2. (I'm not sure about batch-like queries, are there any?)

The use cases I see are mostly around (1). Workflows that I deal with in Bayesian data analysis and clinical trial simulation are arbitrary DAGs of tasks. Many of these DAGs are just composites of map/reduce steps, but this is not always the case, and I would strongly prefer not to make assumptions about the graph topoolgy. 

>And the following gradient of approaches:
>
> 1. Spin up workers once and keep all of them until all computations are complete
> 2. Spin up workers once and shut them down as soon as their work is complete (this is what we are currently using)

Somewhat related: `targets` has the capability of running some targets on the main process and others on remote workers. When it starts traversing the graph, it skips up-to-date targets like always, and it tries to run everything it can locally. It only launches `clustermq` workers when it reaches a target in the graph that (1) is out of date (needs to run) and (2) must to run on a remote worker.

> 4. Spin up a separate worker for each individual function call

(4) The HPC devops team at my work has expressed a strong preference for (4) because it would significantly reduce idling time. For use, our scheduler is beefy, but HPC resources are often occupied. In addition, they claim fully transient workers would allow sys admins to more finely manage what is running on the cluster. However, I am not sure how they would feel about scenarios with large numbers of small jobs (less common at my work). 

In any case, my team and I have not even been able to pilot transient workers (via `tar_make_future()`) because `future.batchtools` is intractably slow. Even if we could, the overhead might be a bit much.

If I recall correctly, `snakemake` used fully transient workers at first, but due to complaints from sys admins, it now supports job groups. This is similar [batching in `targets`](https://books.ropensci.org/targets/dynamic.html#batching), which is recommended in the documentation as part of user best practices. [Target batching](https://books.ropensci.org/targets/dynamic.html#batching) is automatically supported in extension packages such as [`tarchetypes`](https://docs.ropensci.org/tarchetypes/) and [`stantargets`](https://wlandau.github.io/stantargets/). 

> 3. (there is probably some intermediate uses cases here)

What about something that allowed workers to scale up and down as the work progresses?

1. Start by submitting an array job of a certain user-specified size, not necessarily the maximum size.
2. If more work is requested with `$send_call()` and the number of currently busy workers is less than the user-specified maximum, then initialize a new worker for the new job.
3. If a worker idles for long enough (i.e. receives nothing or only `$send_wait()` for some length of time) then shut down that worker.

Tuning the initial worker size in (1) and max idle time in (3) could allow a nice spectrum between fully transient and fully persistent workers.

**Reply by wlandau** [reply]
By the way, as a user, how do I know if all the `clustermq` workers are occupied or some are ready to accept jobs? The [`targets` main loop here](https://github.com/ropensci/targets/blob/62bc91bdd01db47753ec2d6de476ed0c998035a0/R/class_clustermq.R#L163-L174) seems to take care of that, but even though I wrote this code based on your code for the worker API event loop, I do not understand how it accomplishes this. Does it have to do with the value of `message$token`?

**Reply by wlandau** [reply]
In https://github.com/mschubert/clustermq/discussions/257#discussioncomment-567979, what I imagine is that when all the workers are occupied, `message$token` is not equal to "set_common_data_token", which means we call [`self$crew$send_common_data()`](https://github.com/ropensci/targets/blob/62bc91bdd01db47753ec2d6de476ed0c998035a0/R/class_clustermq.R#L166) instead of [`self$next_target()`](https://github.com/ropensci/targets/blob/62bc91bdd01db47753ec2d6de476ed0c998035a0/R/class_clustermq.R#L168). This gives me the impression that common data is sent again again every time we observe that all workers are occupied, but clearly that must not be what is happening. To confirm: does `$send_common_data()` implicitly do nothing if the common data is already sent?

**Reply by mschubert** [reply]
The shutdown of idle workers is already possible with the current API. What I read from your post is that the upscaling is lacking and should be provided (which I completely agree with).

> how do I know if all the `clustermq` workers are occupied

If you get a `WORKER_READY` on the API without having a call lined up, it's idle. I'm currently not tracking the idle time, but that should be straightforward to add. This has nothing to do with the `message$token`.

The common data token is only there to track if a worker already has what is set as `common_data` (which is also what it does in your example, as far as I understand).

As a general note, I'm not quite happy with how this token working and I'm looking to replace is with a worker environment (that is set on the worker object and passed to workers automatically when required).

**Reply by mschubert** [reply]
> The HPC devops team at my work has expressed a strong preference for (4)

I can only see this make sense if all function calls on your HPC take some minutes to compute, and the overall number of calls is relatively low. Otherwise, if some users have a million short calls each this may bring down the whole farm (which is why I will probably never support that).

**Reply by mattwarkentin** [reply]
Again, late to join here, but my opinion is that I like persistent workers capable of handling many jobs. I am definitely looking at `clustermq` through a `targets` lens, but I do think the ability to scale up and down would be great. As you mentioned, scaling down was always supported by `clustermq` and is now working in `targets`, so the ability to scale up as required would be a great feature.

**Reply by mattwarkentin** [reply]
@wlandau I was thinking about this a bit last night, if/when upscaling is supported, I thought about two types of worker management techniques (relevant to `targets`):

1. __Heuristic__ (_less conservative_)
    + This would align with what you described above: (1) start some number of jobs (less than or equal to `max_workers`), (2) spin up more jobs if more work is requested and all current workers are busy, and if `max_workers` isn't saturated, and (3) if any worker is idle for `idle` amount of time, you can shut down that worker. Iterate through (2) and (3) until all the work is done.
    + This approach would be probably be more resource conscious, in that it would free up resources when they are idle and return them to the resource pool for others to use. 

2. __Deterministic__ (_more conservative_)
    + With this approach, steps (1) and (2) would be the same, but you would only shut down workers when those workers would _never_ receive any more work. I think this is basically the current approach in `targets`. Resources are only relinquished when they are definitely no longer needed.

In either approach (or a combination of the two), the number of workers spun up at step (1) would only be as many workers as needed at the very start (up to `max_workers`), and then workers are spun up and shut down according to one of the above methods. Just scribbling my thoughts down.

**Reply by wlandau** [reply]
I like the way you frame those options, @mattwarkentin.

If the heuristic approach also adopts the shutdown rule of the deterministic one, it seems like the latter is a special case of the former with `max_workers == initial_workers` and `idle = Inf`.

**Reply by mattwarkentin** [reply]
> If the heuristic approach also adopts the shutdown rule of the deterministic one, it seems like the latter is a special case of the former with `max_workers == initial_workers` and `idle = Inf`.

Yes, this is exactly what I was thinking too! With the deterministic shutdown rule in place, `idle` becomes like a tuning parameter for the strength of the heuristic rule.

**Comment by wlandau** [comment]
Related: @mschubert, does `clustermq` provide a way to know how many workers are up that have not been sent shutdown messages? To implement https://github.com/ropensci/targets/pull/399, I used a counter in `targets` to keep track of this manually. I am not sure that will still work if/when `clustermq` automatically scales up workers.

**Reply by wlandau** [reply]
I tried `qsys$workers` and `qsys$workers_running` but had trouble. (The event loop was trying to shut down workers that were already shut down.)

**Reply by mschubert** [reply]
Can you elaborate? `w$workers` should give you the correct number.

For worker management in general, the idea is that if you want to shut down a worker, you reply to its `WORKER_READY` message with a `WORKER_STOP` message (in practice, by calling `w$send_shutdown_worker()`). After that, it will still answer with a `WORKER_DONE` and summary statistics, which you do not need to shut down anymore (this is handled internally by the `QSys` base class and you just quit the loop and then call `w$cleanup()`).

**Reply by wlandau** [reply]
That certainly makes sense. I reproduced the issue in a [new branch](https://github.com/ropensci/targets/tree/clustermq-257-worker-counter). If you install https://github.com/ropensci/targets/commit/d2715730de0c86b9f2053b2e499177d6a69ffd03, the following error results in a simple example. (I know the error is slightly different from what I was describing, but it is one of the errors I saw when trying `w$workers` and `w$workers_running`.)

``` r
library(targets)

print(packageDescription("targets")$GithubSHA1)
#> [1] "d2715730de0c86b9f2053b2e499177d6a69ffd03"

tar_script({
  options(clustermq.scheduler = "multicore")
  tar_option_set(deployment = "main")
  list(
    tar_target(index, seq_len(4)),
    tar_target(run, index, pattern = map(index), deployment = "worker"),
    tar_target(end, run)
  )
})

tar_make_clustermq(callr_function = NULL)
#> ● start target index
#> ● built target index
#> ● start branch run_77ba9815
#> ● built branch run_77ba9815
#> ● start branch run_4b7abfa4
#> ● built branch run_4b7abfa4
#> ● start branch run_f5a3cbc2
#> ● built branch run_f5a3cbc2
#> ● start branch run_b0df9eb8
#> ● built branch run_b0df9eb8
#> ● built pattern run
#> Error in self$crew$receive_data(): Trying to receive data after work finished
#> ● end pipeline
```

<sup>Created on 2021-04-11 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

If I use a custom worker counter in `targets` (current implementation) then there is no issue.

``` r
library(targets)

print(packageDescription("targets")$GithubSHA1)
#> [1] "d0422b378eb5ed2900822c89112406ace3e9db7c"

tar_script({
  options(clustermq.scheduler = "multicore")
  tar_option_set(deployment = "main")
  list(
    tar_target(index, seq_len(4)),
    tar_target(run, index, pattern = map(index), deployment = "worker"),
    tar_target(end, run)
  )
})

tar_make_clustermq(callr_function = NULL)
#> ● start target index
#> ● built target index
#> ● start branch run_77ba9815
#> ● built branch run_77ba9815
#> ● start branch run_4b7abfa4
#> ● built branch run_4b7abfa4
#> ● start branch run_f5a3cbc2
#> ● built branch run_f5a3cbc2
#> ● start branch run_b0df9eb8
#> ● built branch run_b0df9eb8
#> ● built pattern run
#> ● start target end
#> ● built target end
#> Master: [0.3s 38.1% CPU]; Worker: [avg 23.7% CPU, max 4159493.0 Mb]
#> ● end pipeline
```

<sup>Created on 2021-04-11 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Reply by mschubert** [reply]
I'll have a look. Given that you have a working solution, I will put this in the low priority bucket.
